src/exo/LoopIR_scheduling.py

```diff
@@ -1493,6 +1504,194 @@ def map_e(self, e):
+                if s.orelse:
+                    check &= check_only_scaled_reduces(s.orelse)
+            elif isinstance(s, LoopIR.Seq):
+                check &= check_only_scaled_reduces(s.body)
+            elif isinstance(s, (LoopIR.WindowStmt, LoopIR.WriteConfig, LoopIR.Call)):
+                raise NotImplementedError(
+                    f"unsupported stmt type in loop body: {type(s)}"
+                )
+            elif isinstance(s, (LoopIR.Pass, LoopIR.Alloc, LoopIR.Free)):
+                pass
+            else:
+                raise NotImplementedError(f"unknown stmt type {type(s)}")
+        return check
+
+    if not check_only_scaled_reduces(loop.body):
+        raise SchedulingError(
+            f"cannot lift constant because the reduces to buffer {assign_s.name} in the loop body are not all of the same form"
+        )
+    if constant is None:
+        raise SchedulingError(
+            "cannot lift constant because did not find a reduce in the loop body of the form `buffer += c * expr`"
+        )
+    if isinstance(constant, LoopIR.Read):
+        for (name, typ) in get_writes_of_stmts(loop.body):
+            if constant.name == name and constant.type == typ:
+                raise SchedulingError(
+                    "cannot lift constant because it is a buffer that is written in the loop body"
+                )
+
+    ir, fwd = assign_cursor.get_root(), lambda x: x
+
+    def lift_constant(stmts):
+        nonlocal ir
+        nonlocal fwd
+        for sc in stmts:
+            s = sc._node
+            if isinstance(s, LoopIR.Reduce):
+                if s.name == assign_s.name and same_write_dest(orig_proc, assign_s, s):
+                    assert (
+                        isinstance(s.rhs, LoopIR.BinOp)
+                        and s.rhs.op == "*"
+                        and isinstance(s.rhs.lhs, (LoopIR.Const, LoopIR.Read))
+                    )
+                    ir, fwd_repl = fwd(sc)._replace([s.update(rhs=s.rhs.rhs)])
+                    fwd = _compose(fwd_repl, fwd)
+            if isinstance(s, LoopIR.Seq):
+                lift_constant(sc.body())
+            if isinstance(s, LoopIR.If):
+                lift_constant(sc.body())
+                if sc.orelse():
+                    lift_constant(sc.orelse())
+
+    lift_constant(loop_cursor.body())
+    ir, fwd_ins = (
+        fwd(loop_cursor)
+        .after()
+        ._insert(
+            [
+                assign_s.update(
+                    rhs=LoopIR.BinOp(
+                        "*",
+                        constant,
+                        LoopIR.Read(
+                            assign_s.name,
+                            assign_s.idx,
+                            assign_s.type,
+                            assign_s.srcinfo,
+                        ),
+                        assign_s.type,
+                        assign_s.srcinfo,
+                    )
+                )
+            ]
+        )
+    )
```

Comment:
Use local variables with descriptive names to avoid this sort of thing The code above is awkward to read.